### PR TITLE
Add container build to CI/CD

### PR DESCRIPTION
**Pull request by lkolacek**
CLOUDBLD-11100

Signed-off-by: Ladislav Kolacek <lkolacek@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- n/a Code coverage from testing does not decrease and new code is covered
- n/a New code has type annotations
- n/a Docs updated (if applicable)
- n/a [Issue](https://github.com/containerbuildsystem/cachi2/issues/13) that tracks breaking changes to the OSBS2 integration is updated (if applicable)

### CODE REVIEW

File: .github/workflows/gating.yaml [hunk 1]
```diff
@@ -101,3 +109,44 @@ jobs:
...
+  build-image:
+    name: Build Cachi2 image and run smoketest on it
+    runs-on: ubuntu-latest
+    steps:
+     # Authenticate to container image registry to push the image
+      - name: Podman Login
+        uses: redhat-actions/podman-login@v1.4
+        with:
+          registry: ${{ env.IMAGE_REGISTRY }}
+          username: ${{ secrets.QUAY_USER }}
```
Comment (lkolacek) [inline .github/workflows/gating.yaml:121]: @chmeliik  @taylormadore  @brunoapimentel  @ejegrova  Anyone knows why I cannot pass repository secrets to GitHub Workflow? :(
* https://github.com/containerbuildsystem/cachi2/actions/runs/3419507522/jobs/5693115442
They were added here:
* https://github.com/containerbuildsystem/cachi2/settings/secrets/actions
Is there any hidden flag to make them "more" public in settings? So far did not find anything.

Comment (chmeliik) [reply .github/workflows/gating.yaml:121]: Could it be that every fork needs its own secrets, same as gitlab?

Comment (chmeliik) [reply .github/workflows/gating.yaml:121]: Now that I'm looking at operator-manifest, it seems that the multiarch-build workflow only happens on merges to master: https://github.com/containerbuildsystem/operator-manifest/actions

Which is what we want, right? We shouldn't be pushing to quay.io for every PR. So it should be OK to have the credentials only in this repo

Comment (lkolacek) [reply .github/workflows/gating.yaml:121]: yes that must be it. I'll add them to my fork to test pushing content and then I will change it to do it only on merges to main/master

**Comment by lkolacek** [comment]
Pushing content to quay.io was checked on PR https://github.com/containerbuildsystem/cachi2/pull/19 (does not work in the forked repo for me ... I need a vacation I would swear it worked yesterday probably swapped my testing PRs)

I checked possibilities of extended smoketest here. Result: smoketest for fetching deps should be enough for CI/CD GitHub actions.
* Generating env file is covered well with unit tests 
* Using Cachi2 resources in container build will be covered in a proper integration test in Jenkins.

File: .github/workflows/gating.yaml [hunk 2]
```diff
@@ -101,3 +110,46 @@ jobs:
...
+      - name: Podman Login
+        uses: redhat-actions/podman-login@v1
+        with:
+          registry: ${{ env.IMAGE_REGISTRY }}
+          username: ${{ secrets.QUAY_USER }}
+          password: ${{ secrets.QUAY_PWD }}
+
+      - name: Push image to Quay.io registry
+        run: |
+          if [[ "${{ github.ref }}" == "refs/heads/main" ]]; then
```
Comment (chmeliik) [inline .github/workflows/gating.yaml:152]: This might be nicer, and useful for podman login https://stackoverflow.com/questions/58139406/only-run-job-on-specific-branch-with-github-actions

Comment (chmeliik) [reply .github/workflows/gating.yaml:152]: Although I'm not sure if you can do the same thing on the "step" level :thinking:

File: .github/workflows/gating.yaml [hunk 3]
```diff
@@ -8,6 +8,15 @@ on:
   workflow_dispatch:
     inputs: {}
 
+env:
+  IMAGE_NAME: cachi2
+  IMAGE_REGISTRY: quay.io
+  IMAGE_NAMESPACE: containerbuildsystem
+  IMAGE_TAG: latest
```
Comment (MartinBasti) [inline .github/workflows/gating.yaml:15]: should we differentiate between PR and main branch builds?

File: .github/workflows/gating.yaml [hunk 4]
```diff
@@ -101,3 +110,46 @@ jobs:
         ignore: DL3041
         failure-threshold: warning
 
+  build-image:
+    name: Build Cachi2 image and run smoketest on it
+    runs-on: ubuntu-latest
+    steps:
+      - name: add checkout action...
+        uses: actions/checkout@v3
```
Comment (chmeliik) [inline .github/workflows/gating.yaml:118]: How come we don't need a full clone? :thinking:  Does cachi2 in the built image have the correct version? (`cachi2 --version`)

```yaml
- uses: actions/checkout@v3
  with:
    fetch-depth: 0
```

File: .github/workflows/gating.yaml (hunk 3 above)
Comment (lkolacek) [reply .github/workflows/gating.yaml:15]: Not sure if I get this comment. Currently, we build images locally in open PR, but we do not push them anywhere. Only merged PRs will update the cachi2:latest image in quay.io . Can you please elaborate more here?

Comment (MartinBasti) [reply .github/workflows/gating.yaml:15]: ah sorry, I haven't noticed the condition in task

**Comment by taylormadore** [comment]
Looks great! This is probably more a question for later when we add the integration tests job to the CI, but have you considered how we're going to share the built images between jobs if we're only pushing to quay when running on main? IIRC in cachito CI the image builds and integration tests both happen within the same job

File: .github/workflows/gating.yaml (hunk 4 above)
Comment (lkolacek) [reply .github/workflows/gating.yaml:118]: Fixed!

File: .github/workflows/gating.yaml (hunk 2 above)
Comment (lkolacek) [reply .github/workflows/gating.yaml]: Looks like it works. Thanks for the tip adam!

**Comment by lkolacek** [comment]
> Looks great! This is probably more a question for later when we add the integration tests job to the CI, but have you considered how we're going to share the built images between jobs if we're only pushing to quay when running on main? IIRC in cachito CI the image builds and integration tests both happen within the same job

Thanks for question Taylor. 
I believe it will work almost the same way as in Cachito. We will push new cachi2 changes via GitHub PR action pipeline to quay.io - quay.io/containerbuildsystem/cachi2:latest .. Image on latest tag will get updated and IIRC based on this change webhook will trigger integration tests in Jenkins. If that won't be possible we will simply run nightly tests as always which will automatically use latest cachi2 image.   


I addressed comments.

Tested CI here also with pushing content to Quay.io
* https://github.com/containerbuildsystem/cachi2/pull/19

**Comment by chmeliik** [comment]
> Image on latest tag will get updated and IIRC based on this change webhook will trigger integration tests in Jenkins.

I was hoping integration tests would run upstream, same as they do in Cachito 1. And since Cachi2 won't be a service, I was hoping there would be no Jenkins tests at all

**Comment by lkolacek** [comment]
> > Image on latest tag will get updated and IIRC based on this change webhook will trigger integration tests in Jenkins.
> 
> I was hoping integration tests would run upstream, same as they do in Cachito 1. And since Cachi2 won't be a service, I was hoping there would be no Jenkins tests at all

Yes, upstream tests are planned. I am wondering only about two merged changes approximately at the same time which would pass integration tests but break workflow together -> but that would be corner case, second PR would probably fail to merge on conflict. Anyway this will be resolved in integration tests design doc.

File: .github/workflows/gating.yaml [hunk 5]
```diff
@@ -8,6 +8,15 @@ on:
...
     inputs: {}
 
+env:
+  IMAGE_NAME: cachi2
+  IMAGE_REGISTRY: quay.io
+  IMAGE_NAMESPACE: containerbuildsystem
+  IMAGE_TAG: latest
+  TEST_REPO: https://github.com/release-engineering/retrodep.git
+  TEST_BRANCH: v2.1.1
+  TEST_FOLDER: /tmp/source
```
Comment (chmeliik) [inline .github/workflows/gating.yaml:18]: Might just be me, but I prefer fewer variables with a clearer purpose

The TEST variables don't feel super necessary, since they are used only within the single smoketest step. When we add integration tests, they might even become confusing

Not sure about the IMAGE variables, they make more sense than the test ones but I kinda prefer having the whole pullspec in one place to make it `grep`-able

File: .github/workflows/gating.yaml [hunk 6]
```diff
@@ -101,3 +110,50 @@ jobs:
...
+        with:
+          registry: ${{ env.IMAGE_REGISTRY }}
+          username: ${{ secrets.QUAY_USER }}
+          password: ${{ secrets.QUAY_PWD }}
+
+      - name: Push image to Quay.io registry
+        if: github.ref == 'refs/heads/main'
+        run: |
+          podman push ${{ env.IMAGE_NAME }}:${{ github.sha }} \
+            ${{ env.IMAGE_REGISTRY }}/${{ env.IMAGE_NAMESPACE }}/${{ env.IMAGE_NAME }}:${{ env.IMAGE_TAG }}
```
Comment (chmeliik) [inline .github/workflows/gating.yaml:159]: How are we  going to preserve multiple versions in the registry?

Should we push both
```
quay.io/containerbuildsystem/cachi2:ef0c82bb199a56e8a9ff49b2bbe7dfc0cf00efc1
quay.io/containerbuildsystem/cachi2:latest
```
To give users a chance to pin to a specific commit?

And/or should we do something like operator-manifest, which also [builds on release](https://github.com/containerbuildsystem/operator-manifest/blob/51425a5327fe771e134a97f59183eb652dcb2459/.github/workflows/multiarch-build.yaml#L7) and [pushes the release tag](https://github.com/containerbuildsystem/operator-manifest/blob/51425a5327fe771e134a97f59183eb652dcb2459/.github/workflows/multiarch-build.yaml#L78)?

File: .github/workflows/gating.yaml (hunk 5 above)
Comment (lkolacek) [reply .github/workflows/gating.yaml:18]: +1 for making pullspec `grep`-able .. I am ok with removing vars completely if you don't mind changing these vars on many places during refactoring

File: .github/workflows/gating.yaml (hunk 6 above)
Comment (lkolacek) [reply .github/workflows/gating.yaml]: That make sense, let's push both. Might be useful for RCA and debugging.

I haven't thought about releases yet but it does not make sense to wait for anything. +1 I'll add it now. I hope it works the same way. :)

File: .github/workflows/gating.yaml (hunk 5 above)
Comment (chmeliik) [reply .github/workflows/gating.yaml:21]: I personally value readability over refactor-ability in CI definitions. Otherwise I end up having to mentally substitute the variables when reading the tasks, and refactoring is usually a search+replace operation for me anyway :smile: 

But I'd like others' opinions on this as well

File: .github/workflows/gating.yaml (hunk 6 above)
Comment (chmeliik) [reply .github/workflows/gating.yaml]: I'm thinking whether we want to run the whole pipeline (tests, linters, everything) on release or just the image build. But it probably makes sense to run the whole thing

Comment (lkolacek) [reply .github/workflows/gating.yaml]: I would say we want to run the whole thing when it comes to release

File: .github/workflows/gating.yaml (hunk 5 above)
Comment (lkolacek) [reply .github/workflows/gating.yaml:21]: if no one will object I'll remove variables :)

File: .github/workflows/gating.yaml (hunk 6 above)
Comment (lkolacek) [reply .github/workflows/gating.yaml]: but I can prepare a separate workflow just for the build and testing as operator-manifest has it https://github.com/containerbuildsystem/operator-manifest/blob/51425a5327fe771e134a97f59183eb652dcb2459/.github/workflows/multiarch-build.yaml

File: .github/workflows/gating.yaml (hunk 5 above)
Comment (taylormadore) [reply .github/workflows/gating.yaml:21]: +1

File: .github/workflows/gating.yaml [hunk 7]
```diff
@@ -101,3 +113,57 @@ jobs:
...
+      - name: Podman Login
+        uses: redhat-actions/podman-login@v1
+        if: github.ref == 'refs/heads/main'
+        with:
+          registry: ${{ env.IMAGE_REGISTRY }}
+          username: ${{ secrets.QUAY_USER }}
+          password: ${{ secrets.QUAY_PWD }}
+
+      - name: Push image to Quay.io registry
+        if: github.ref == 'refs/heads/main'
```
Comment (chmeliik) [inline .github/workflows/gating.yaml:159]: I think this won't run for a [release event](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release). We could probably add something like "or event == release", but maybe we could come back to release workflows later.

Maybe releasing should just be taking `quay.io/containerbuildsystem/cachi2:${{ github.sha }}` and adding the current tag to it, maybe we don't need to build a whole new image

Comment (chmeliik) [reply .github/workflows/gating.yaml:159]: Btw. @brunoapimentel do you have ideas how versioning of the cachi2 image will be done in HACBS? Should we have tags for each released semantic version? Are just the commit-hash tags good enough? https://github.com/redhat-appstudio/build-definitions/pull/218#discussion_r1011502124

Comment (brunoapimentel) [reply .github/workflows/gating.yaml:159]: The commit hash would be enough for HACBS, but we could also use semantic versions. In the latter case, I would also add the image sha, this way we'd make sure we're pinning a specific image version to the pipeline.

The image versions are being bumped manually in HACBS for now. I have filed a ticket so we can decide on a process on how to update the image tags there.

Comment (chmeliik) [reply .github/workflows/gating.yaml:159]: Image sha as in manifest digest? Pinning to digest is the universal way to pin a specific image
```
quay.io/containerbuildsystem/cachi2@sha256:bc87b6bd5549b4b4cb5d459a5edaa0e705e2ac74d0ae7fcdb029cac3da82be76
```

Or image sha as in commit sha? I think we should push commit shas on merge to main and tag those shas with the semver tag on release

File: .github/workflows/gating.yaml (hunk 5 above)
Comment (lkolacek) [reply .github/workflows/gating.yaml]: vars removed

File: .github/workflows/gating.yaml (hunk 7 above)
Comment (lkolacek) [reply .github/workflows/gating.yaml:150]: +1 for adding also image sha, do you want to handle releases here? or in a separate PR?

Comment (chmeliik) [reply .github/workflows/gating.yaml:150]: I'd vote for a separate PR with a tiny design first

Comment (lkolacek) [reply .github/workflows/gating.yaml:147]: I dropped part for releases for now and filed ticket CLOUDBLD-12459 for design and investigation. Any more comments for this PR?

**Review by chmeliik — APPROVED** [review]
LGTM :rocket:

**Comment by chmeliik** [comment]
Minor detail: commit message still mentions releases

**Comment by chmeliik** [comment]
Docs are probably good enough: https://github.com/containerbuildsystem/cachi2#installation

**Review by taylormadore — APPROVED** [review]
LGTM :+1:

**Review by ejegrova — APPROVED** [review]
LGTM
